Ullaakut/Bloggo
# Add jwt token parsing and access service unit tests with real tokens

## Goal of this PR

Add parsing of JWT to the middleware. The previous version just checks if the token is a known user ID, while here we parse the JWT, validate it, and extract the user ID from it.

For now, the known user IDs are only stored in memory and not yet in a database though.

Fixes #11 

## How to test it

You can use [jwt.io](https://jwt.io/) to edit the valid token that can be found in the `README.md` file at the root of the repository. Try to trigger errors in different ways:

* Invalid token syntax (not respecting the x.y.z format for example or having invalid characters)
* Invalid claim format (using strings instead of integers and vice-versa)
* User not found (simply changing the value of the `sub` claim to another string should trigger this error)
* Token issued in the future
* Token expired
* Token source is not trusted (simply changing the value of the `iss` claim to another string should trigger this error)

Each one of these errors should return a `401 Unauthorized` response, and a body like such:

```
{"message":"could not validate token: Token is expired"}
```

Of course, also make sure to try using the provided token in the `README.md` file at the root of the repository, and make sure that you get a `200 OK` response

To make HTTP requests to the server, you can use postman or use a curl request like such:

```
curl -X POST \
       http://0.0.0.0:4242/posts \
       -H 'Authorization: Bearer eyJ0eXAiOiJKV1QiLCJhbGciOiJIUzI1NiJ9.eyJlbWFpbCI6ImJyZW5kYW4ubGUtZ2xhdW5lYyt0ZXN0YXBpQGVwaXRlY2guZXUiLCJlbWFpbF92ZXJpZmllZCI6ZmFsc2UsImlzcyI6Imh0dHBzOi8vc2FtcGxlcy5hdXRoMC5jb20vIiwic3ViIjoiYXV0aDB8NTk2ZjI3YzJjMzcwOTY2MWU5Y2VhMzdkIiwiYXVkIjoia2J5dUZEaWRMTG0yODBMSXdWRmlhek9xak8zdHk4S0giLCJleHAiOjE2MDA0OTI5NjUsImlhdCI6MTUwMDQ1Njk2NX0.4To8mYgu4pM7J2G5jBTnKWelCTU1U1jo0QOENVp3pOk' \
       -H 'Cache-Control: no-cache' \
       -H 'Postman-Token: ddf103dc-3ba5-4f08-adc3-6b9625b7581b'
```

<img width="957" alt="screenshot 2018-08-02 at 19 51 04" src="https://user-images.githubusercontent.com/6976628/43601428-69adba66-968d-11e8-93e4-297c1157ab0f.png">
